osbuild/qemu.py

```diff
@@ -0,0 +1,328 @@
+        self,
+        share_dir: str,
+        socket_path: str,
+        readonly: bool,
+    ):
+        self.share_dir = share_dir
+        self.cmd = [
+            "/usr/libexec/virtiofsd",
+            "--socket-path=" + socket_path,
+            "--shared-dir", share_dir,
+            "--cache", "always",
+            "--sandbox=none",
+            "--xattr",
+            "--security-label",
+        ]
+        if readonly:
+            self.cmd += ["--readonly"]
+        self._proc: Optional[subprocess.Popen] = None
+
+    def start(self) -> int:
+        if self._proc is not None:
+            return self._proc.pid
+
+        self._proc = subprocess.Popen(
+            self.cmd,
+            text=False,
+            close_fds=True,
+        )
+
+        return self._proc.pid
+
+    def stop(self) -> None:
+        p = self._proc
+        if not p:
+            return
+
+        try:
+            p.terminate()
+        except ProcessLookupError:
+            self._proc = None
+            return
+
+        # Wait a bit for graceful exit
+        try:
+            p.wait(timeout=5.0)
+        except subprocess.TimeoutExpired:
+            try:
+                p.kill()
+            except ProcessLookupError:
+                pass
+            try:
+                p.wait(timeout=2.0)
+            except subprocess.TimeoutExpired:
+                pass
+
+        self._proc = None
+
+    # Context manager
+    def __enter__(self):
+        self.start()
+        return self
+
+    def __exit__(self, exc_type, exc, tb):
+        self.stop()
+        return False
+
+
+def find_qemu():
+    arch = platform.machine()
+    binary_names = [f"qemu-system-{arch}", "qemu-kvm"]
+
+    for binary_name in binary_names:
+        qemu_bin_dirs = ["/usr/bin", "/usr/libexec"]
+        if "PATH" in os.environ:
+            qemu_bin_dirs += os.environ["PATH"].split(":")
+
+        for d in qemu_bin_dirs:
+            p = os.path.join(d, binary_name)
+            if os.path.isfile(p):
+                return p
+
+    raise RuntimeError("Can't find qemu binary")
+
+
+def qemu_available_accels(qemu):
+    info = subprocess.check_output([qemu, "-accel", "help"]).decode('utf-8').splitlines()
+    # First line is header, skip it
+    return info[1:]
+
+
+class Qemu:
+    def __init__(
+        self,
+        mem: str,
+        kernel_path: str,
+        initrd_path: str,
+        rootfs_path: str,
+        libdir_path: str,
+        serial_stdout: bool = False,
+    ) -> None:
+        self._pid: Optional[int] = None
+
+        self._tmpdir = tempfile.TemporaryDirectory(prefix="osbuild-qemu-")
+        self.pidfile = os.path.join(self._tmpdir.name, "qemu.pid")
+        self.serials: Dict[str, str] = {}
+        self.virtiofs: Dict[str, Tuple[str, Virtiofsd]] = {}
+
+        qemu_bin = find_qemu()
+        qemu_accels = qemu_available_accels(qemu_bin)
+
+        self.cmd = [
+            qemu_bin,
+            "-daemonize",
+            "-pidfile",
```

Comment:
(super nitpick) I wonder if it would be slightly nicer to group the matching options/args together, ie.e.
```python
 self.cmd = [
            qemu_bin,
            "-daemonize",
            "-pidfile", self.pidfile,
            "-display", "none",
...
```